src/Storages/StorageKeeperMap.cpp

```diff
@@ -0,0 +1,329 @@
+{
+    const auto * function = elem.as<ASTFunction>();
+    if (!function)
+        return std::nullopt;
+
+    if (function->name != "equals")
+        return std::nullopt;
+
+    const auto & args = function->arguments->as<ASTExpressionList &>();
+    const ASTIdentifier * ident;
+    ASTPtr value;
+    if ((ident = args.children.at(0)->as<ASTIdentifier>()))
+        value = args.children.at(1);
+    else if ((ident = args.children.at(1)->as<ASTIdentifier>()))
+        value = args.children.at(0);
+    else
+        return std::nullopt;
+
+    if (ident->name() != "key")
+        return std::nullopt;
+
+    auto evaluated = evaluateConstantExpressionAsLiteral(value, context);
+    const auto * literal = evaluated->as<ASTLiteral>();
+    if (!literal)
+        return std::nullopt;
+
+    if (literal->value.getType() != Field::Types::String)
+        return std::nullopt;
+
+    return KeyFilter{literal->value.safeGet<std::string>(), FilterType::EXACT};
+}
+
+Pipe StorageKeeperMap::read(
+    const Names & column_names,
+    const StorageSnapshotPtr & /*storage_snapshot*/,
+    SelectQueryInfo & query_info,
+    ContextPtr context,
+    QueryProcessingStage::Enum /*processed_stage*/,
+    size_t max_block_size,
+    unsigned /*num_streams*/)
+{
+    Block sample_block;
+    for (const std::string & column_name : column_names)
+    {
+        sample_block.insert({std::make_shared<DataTypeString>(), column_name});
+    }
+
+    const auto & select = query_info.query->as<ASTSelectQuery &>();
+    std::optional<KeyFilter> key_filter;
+    if (select.where())
+        key_filter = tryGetKeyFilter(*select.where(), context);
+
+    return Pipe(std::make_shared<StorageKeeperMapSource>(sample_block, *this, max_block_size, std::move(key_filter)));
+}
+
+SinkToStoragePtr StorageKeeperMap::write(const ASTPtr & /*query*/, const StorageMetadataPtr & /*metadata_snapshot*/, ContextPtr /*context*/)
+{
+    auto columns = getNamesAndTypes();
+    Block write_header;
+    for (const auto & [name, type] : columns)
+    {
+        write_header.insert(ColumnWithTypeAndName(type, name));
+    }
+
+    return std::make_shared<StorageKeeperMapSink>(write_header, *this);
+}
+
+zkutil::ZooKeeperPtr & StorageKeeperMap::getClient()
+{
+    if (zookeeper_client->expired())
+        zookeeper_client = zookeeper_client->startNewSession();
+
+    return zookeeper_client;
+}
+
+const std::string & StorageKeeperMap::rootKeeperPath() const
+{
+    return keeper_path;
+}
+
+void registerStorageKeeperMap(StorageFactory & factory)
+{
+    factory.registerStorage(
+        "KeeperMap",
+        [](const StorageFactory::Arguments & args)
+        {
+            if (!args.attach && !args.columns.empty())
+                throw Exception(
+                    ErrorCodes::BAD_ARGUMENTS,
+                    "Storage KeeperMap does not accept column definition as it has predefined columns (key String, value String)");
+
+            ASTs & engine_args = args.engine_args;
```

Comment:
I'd like to have something like a limit on the total number of keys (it can be easily checked by loading numChildren at startup and updating the counter).